### ISSUE TITLE
Edge case in 'array-bracket-newline' and consistency towards 'object-curly-newline'

### ISSUE THREAD

**Issue by ASDFGerte** [eslint/eslint#14001]
**What rule do you want to change?** `array-bracket-newline`
**Does this change cause the rule to produce more or fewer warnings?** Depends on settings
**How will the change be implemented?** Change of how options work, or a new option.

My desired style always allows putting newlines on brackets. Brackets always need to be consistent. Newlines are required, if there is any newline inside of the array (between elements, or in an element, the latter being another problem, related to `array-element-newline` being different than `object-property-newline`, but this will be mostly disregarded here). Single-line arrays are allowed. The main problem case is a single element. The following should be valid:
```js
let v = [
  0,
];
```
As `multiline` doesn't see any newlines for elements, it disallows the newlines for the brackets. `consistent` doesn't work either, as it would allow to have no newlines for the brackets, when there are newlines between (or in) elements.

---

I see some rules as related, e.g. `array-bracket-newline` and `object-curly-newline`, or `array-element-newline` and `object-property-newline`. While surely not identical, they cover very similar areas.

Above behavior is best learned by reading the docs of `object-curly-newline`:
> "multiline": true requires line breaks if there are line breaks inside properties or between properties. **Otherwise, it disallows line breaks**.

The emphasized part also applies for `array-bracket-newline`, although not explicitly stated there. I don't fully understand, why this was added in the first place, or why it is mentioned in one rule, but not on the other.

Apart from my desired behavior not working, there is an inconsistency: `object-curly-newline`, which as mentioned should imho be very similar, has different behavior. There, `mutliline` and `consistent` are both options. Through what looks to me like an unintended loophole (it's not mentioned anywhere, and looks to be a side effect of the specific implementation), it's possible to set both `mutliline` and `consistent` at the same time, solving the issue. If you ask me, these rules should offer nearly identical options, and this behavior should be mentioned somewhere.

I hope the following is unnecessary, but why is this style even desired? Mainly, because some arrays (and objects) frequently get items added or removed. This should be easy, and produce 'nice' changes in version control. An array may temporarily end up with only one element, but changing it to single-line is not the intention, as it could soon change back (adding useless edits, and less readable commits). Having single-line arrays should at the same time not be forbidden.

---

This can be solved in a variety of ways, including, but probably not limited to:

 - remove the added behavior of 'Otherwise, it disallows line breaks', so `multiline` doesn't forbid newlines for brackets (though there probably was a reason to introduce this, which i just don't know of)
 - add a new option to disable this behavior
 - 'elevate' the `consistent` string option to a property on the options, and allow the same loophole from `object-curly-newline` (with documentation - even in `object-curly-newline` there already is a mention for `consistent` changing the behavior of `min-items`, this would be similar)
 - Do a slightly larger change to the options, and come up with a potentially more intuitive naming and set of options, rather than having one option change the behavior of several others

I'd prefer it, if the options were analogous to `object-curly-newline`. I am however aware, that breaking changes can be very hard to do. Overhauling the options/documentation of both may be the more proper solution, but simply adding a new option just to `array-bracket-newline` would be very easy.

---

PS: There are other minor related issues, e.g. that `multiline` for `array-element-newline` only considers newlines inside elements, but not between, which is different to what `multiline` means in the above. Similarly, `consistent` is a string option there, making these mutually exclusive, while forcing newlines may be desired, when there are any newlines in the array (actually is for me...). Addressing these (and others) together might be a good idea. I am unsure, whether this is asking too much, and I should rather take it one step at a time.

---

**Are you willing to submit a pull request to implement this change?**
If nothing is done, I'll probably create a fork of the rule, make some quick&dirty change to accommodate for my needs, and be done. The problem here is finding a 'proper' solution, which I'll probably not make, or even be in a position to do myself. For me, 'proper' means intuitive and robust, yet trying to avoid breaking changes as much as possible. In a custom fork for personal use, I'd have to consider none of that.

**Comment by ASDFGerte** [comment]
Unlucky. Had searched before, but not found anything, then typed this, but only afterwards found https://github.com/eslint/eslint/issues/13582 which is basically the same thing. Guess i got to fork and change the rule myself. It's a bit sad, because as the author of that issue mentions, it would be better, if there was a way, where not everyone had to implement the same custom rule over and over. I bet he has one, probably others do too, and now I'll have to make yet another.
